File: src/matlab/dft_fconv2.m
function [y] = dft_fconv2(x0, x1)
% Computation of fast convolution with N/2-point DFT (type 2)
% Input:
%	x0 - N/2-point complex-valued sequence 
%		x0 = DFT[ s(2k-1)+i*s(2k) ], where s is N-point real-valued sequence
%	x1 - N/2-point complex-valued sequence of impulse response after DFT
%		x1 = DFT[ m(2k-1)+i*m(2k) ], where m is N-point real-valued sequence
% Output:
%	y - N/2-point complex-valued sequence, y = x0*x1
N=numel(x0)*2;

w0=zeros(1,N/2);
w1=zeros(1,N/2);

h0=zeros(1,N/2);
h1=zeros(1,N/2);

y=zeros(1,N/2);

for k=1:N/2
	w0(k)=0.25*(3-exp(-4i*pi*(k-1)/N));
	w1(k)=0.25*(1+exp(-4i*pi*(k-1)/N));
end

x1(N/2+1)=x1(1);
for k=1:N/2
	h0(k)=w0(k)*x1(k) + w1(k)*conj(x1(N/2-k+2));
	h1(k)=w1(k)*(x1(k) - conj(x1(N/2-k+2)));
end

x0(N/2+1)=x0(1);
for k=1:N/2
	y(k)= x0(k)*h0(k) + conj(x0(N/2-k+2))*h1(k);
end

end

File: src/dft_fconv2.c
/**
 * @file dft_fconv2.c
 * @brief Computation of fast convolution with N/2-point DFT (type 2)
 * @author matyunin.d
 * @date 29.08.2019
 */

#include <complex.h>
#include <math.h>

/**
 * @brief Computation of fast convolution with N/2-point DFT (type 2)
 * @param x Pointer to N/2-point complex-valued sequence, x = DFT[ s(2k)+i*s(2k+1) ], s - N-point real-valued sequence
 * @param h Pointer to N/2-point complex-valued sequence, h = DFT[ m(2k)+i*m(2k+1) ], s - N-point real-valued sequence (impulse response)
 * @param y Pointer to output N/2-point complex-valued sequence
 * @param N Length
 */
void dft_fconv2(const double complex *x, const double complex *h, double complex *y, unsigned int N)
{
	double complex w0, w1;
	double complex h0, h1;

	y[0] = x[0] * creal(h[0]);

	for (unsigned int k = 1; k < N/2; k++) {
		w0 = 0.25 * (3.0 - cexp(-4.0 * I * M_PI * k / N));
		w1 = 0.25 * (1.0 + cexp(-4.0 * I * M_PI * k / N));

		h0 = w0 * h[k] + w1 * conj(h[N/2 - k]);
		h1 = w1 * (h[k] - conj(h[N/2 - k]));

		y[k] = x[k] * h0 + conj(x[N/2 - k]) * h1;
	}
}


